### PR TITLE
Feature/issue 855 log inv logit diff

### PR DESCRIPTION
**Pull request by andrjohns**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Address #855, adding a standalone function (```log_inv_logit_diff```) with gradients for the log of the difference between two inverse logits. This replaces the function declared in ```prim/mat/prob/ordered_logistic_lpmf```.

#### Intended Effect:
Improve the speed and efficiency of calculating the log difference between two inverse logits.

#### How to Verify:
Tests are included for prim/rev/fwd/mix

#### Side Effects:
N/A

#### Documentation:
As doxygen

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Andrew Johnson

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/fwd/scal/fun/log_inv_logit_diff.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include <stan/math/fwd/core.hpp>
+#include <stan/math/prim/scal/fun/log_inv_logit_diff.hpp>
+#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/scal/fun/inv.hpp>
+#include <stan/math/prim/scal/fun/expm1.hpp>
+
+namespace stan {
+namespace math {
+
+using std::exp;
```
Comment (bob-carpenter) [inline stan/math/fwd/scal/fun/log_inv_logit_diff.hpp]: No using statements at the top-level of namespaces.  

Could you please check that others haven't been added elsewhere?  I've seen a few of these in recent PRs.

File: stan/math/fwd/scal/fun/log_inv_logit_diff.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+template <typename T>
```
Comment (bob-carpenter) [inline stan/math/fwd/scal/fun/log_inv_logit_diff.hpp]: These should all have doc as they're standalone functions.  It can be simple, but should indicate tparams, params, and return by name.

File: stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp [hunk 3]
```diff
@@ -80,7 +75,7 @@ typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
   } else if (y == K) {
     logp_n -= log1p_exp(c[K - 2] - lambda);
   } else {
-    logp_n += log_inv_logit_diff(c[y - 2] - lambda, c[y - 1] - lambda);
+    logp_n += log_inv_logit_diff(lambda - c[y - 2], lambda - c[y - 1]);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp]: Why are these changing orders?  It changes the sign of the arguments.  Dos it not matter for this function or is it a different function?

File: stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp [hunk 4]
```diff
@@ -146,8 +141,8 @@ typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
     } else if (y[i] == K) {
       logp_n -= log1p_exp(c[K - 2] - lambda[i]);
     } else {
-      logp_n += log_inv_logit_diff(c[y[i] - 2] - lambda[i],
-                                   c[y[i] - 1] - lambda[i]);
+      logp_n += log_inv_logit_diff(lambda[i] - c[y[i] - 2],
```
Comment (bob-carpenter) [inline stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp]: same question here.  at least it seems to be systematic!

File: stan/math/prim/scal/fun/log_inv_logit_diff.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <stan/math/prim/scal/fun/log1m_exp.hpp>
+#include <stan/math/prim/scal/fun/log1p_exp.hpp>
+#include <boost/math/tools/promotion.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the natural logarithm of the difference of the
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/log_inv_logit_diff.hpp]: Perfect.  Thanks for documenting this so carefully.

File: stan/math/prim/scal/fun/log_inv_logit_diff.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * inverse logits of the specified arguments.
+ *
+   \f[
+     \mbox{log_inv_logit_diff}(x,y) =
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/log_inv_logit_diff.hpp]: This won't work in LaTeX without escaping `_` as `\_`.  

And `\mathrm{}` is better as it scales with its environments' size.  So that's `\mathrm{log\_inv\_logit\_diff}`.

File: stan/math/prim/scal/fun/log_inv_logit_diff.hpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * @tparam T1 Type of x argument.
+ * @tparam T2 Type of y argument.
+ * @param x Argument.
+ * @param y Argument.
+ * @return Result of log difference inverse logits of arguments.
+ */
+template <typename T1, typename T2>
+inline typename boost::math::tools::promote_args<T1, T2>::type
+log_inv_logit_diff(const T1& x, const T2& y) {
+  return x + log1m_exp(y - x) - log1p_exp(y) - log1p_exp(x);
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/log_inv_logit_diff.hpp]: I like this implementation.  

[optional]
One thing you could do to make it a bit tighter in the face of autodiff is to group the like types together, e.g.,

```
return x + log1p_exp(x) + ...;
```

Everything's left associative and that's guaranteed not to require any undue promotion.  The rest of the order won't matter.

File: stan/math/rev/scal/fun/log_inv_logit_diff.hpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/scal/fun/log_inv_logit_diff.hpp>
+#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/scal/fun/inv.hpp>
+#include <stan/math/prim/scal/fun/expm1.hpp>
+
+namespace stan {
+namespace math {
+using std::exp;
```
Comment (bob-carpenter) [inline stan/math/rev/scal/fun/log_inv_logit_diff.hpp]: remove namespace-level using

File: test/unit/math/fwd/scal/fun/log_inv_logit_diff_test.cpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#include <stan/math/fwd/scal.hpp>
```
Comment (bob-carpenter) [inline test/unit/math/fwd/scal/fun/log_inv_logit_diff_test.cpp]: Thanks for writing all these tests out.  I think going forward we're going to be OK with just the rev and fwd tests without the additional mix tests.  Please leave them in now that you did them.  The whole thing is going to have to get replaced with an autodiff test framework.

File: stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp (hunk 3 above)
Comment (andrjohns) [reply stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp]: Right, I should have made a note here. The previous (inline) implementation of the function evaluated to:
```
log_inv_logit_diff(x,y) = log(inv_logit(y) - inv_logit(x))
```
Which isn't consistent with other two-argument functions (like ```log_diff_exp```). The new (standalone) function evaluates to:
```
log_inv_logit_diff(x,y) = log(inv_logit(x) - inv_logit(y))
```
So I had to flip the sign of the arguments to keep the result the same.

File: stan/math/fwd/scal/fun/log_inv_logit_diff.hpp (hunk 1 above)
Comment (andrjohns) [reply stan/math/fwd/scal/fun/log_inv_logit_diff.hpp]: Ah sorry about that, I've definitely had them in some tests that I've submitted before. I'll do some searching and submit a pr to remove them.

**Comment by bob-carpenter** [comment]
@andrjohns  Is this ready for final approval?

**Comment by andrjohns** [comment]
Yes, changes are all in and passing tests so should be good to go
________________________________
